### PR TITLE
Added a method to the phantom object to set the logging level.

### PR DESCRIPTION
**Pull request by etienne51**
### Proposed changes in this pull request

I needed a way to set the logging level for this library, to be able to hide all messages with levels higher than 'warn'. So I added the simple 'setLogLevel' method to the 'phantom' object, and its related documentation. I hope it's okay with you. Thanks!

@amir20 to review

### CODE REVIEW

File: src/phantom.js [hunk 1]
```diff
@@ -107,6 +107,15 @@ export default class Phantom {
      * Returns a value in the global space of phantom process
```
Comment (amir20) [inline src/phantom.js]: Update the doc to reflect the method

File: src/phantom.js [hunk 2]
```diff
@@ -107,6 +107,15 @@ export default class Phantom {
...
      * @returns {Promise}
      */
+    setLogLevel(level) {
+        if (typeof level === 'string')
```
Comment (amir20) [inline src/phantom.js]: npm test will fail

**Comment by amir20** [comment]
Broken build

File: src/phantom.js [hunk 3]
```diff
@@ -107,6 +107,15 @@ export default class Phantom {
      * Returns a value in the global space of phantom process
      * @returns {Promise}
      */
```
Comment (amir20) [inline src/phantom.js]: Maybe a better option would be to do 

```js
this.logger = logger
``` 
in the constructor so that others can do whatever they want.

Comment (etienne51) [inline src/phantom.js]: I'm sorry, indeed, I changed the readme file, but I completely forgot the
doc this time...

Yes, that would be a good idea I think! So people could use their own
external loggers for more customisation.

2016-08-20 0:31 GMT+04:00 Amir Raminfar <notifications@github.com>:

**Comment by Dagrut** [comment]
Hi @amir20 !

I'm working with @etienne51 .

We didn't want to make too important changes, that's why we only added a method to change the verbosity.

About your suggestion to provide the logger to everyone, this would make a local variable reachable through multiple instances, which seems weird... Maybe would it be better to make one logger instance per Phantom class? Would you be okay with it?

About other problems (tests, travis, ...), we'll take care of it :-) .
But tomorrow, it's late here!

Thank you for your quick feedback (and also for this nice module!)

File: src/phantom.js [hunk 4]
```diff
@@ -69,41 +70,49 @@ export default class Phantom {
...
         });
 
         this.heartBeatId = setInterval(this._heartBeat.bind(this), 100);
     }
 
     /**
+     * Returns the winston logger instance used to log messages
+     * @returns {winston}
+     */
+    getLogger() {
```
Comment (amir20) [inline src/phantom.js]: You don't need this anymore. Since `this.logger` already exposes the logger as public

File: README.md [hunk 5]
```diff
@@ -103,6 +103,22 @@ It may be a good idea to register handlers to `SIGTERM` and `SIGINT` signals wit
...
+### `phantom#getLogger`
+
+Returns the [winston](https://www.npmjs.com/package/winston) logger instance used to log information to `stdout`. The returned logger is local to its `phantom` instance.
+
+It allows you to increase or decrease the verbosity, redirect messages to a file, and so on.
+
+Also note that the default logging level is `info` and can be set to `debug` by setting the `DEBUG` environment variable to `"true"`.
+
+```js
+var phantom = require('phantom');
```
Comment (amir20) [inline README.md]: This is now outdated and not accurate?

**Comment by amir20** [comment]
Hey @Dagrut no problem. PRs will take some collaboration between us until we get it right. For now, I think you can do a couple of things. You can provide a logger to `create()` method and default to the one that is already provided. That makes more sense because right now there is no way to change the log level until after an instance has been created, which sounds broken. 

I also like allowing logger to be provided to the constructor because then people can choose to use their own implementation of logger, which sounds better. 

Thoughts?

**Comment by amir20** [comment]
@elias-winberg fixed the broken builds and I just merged it.

**Comment by Dagrut** [comment]
@amir20 Okay!

Since all your object properties are undocumented, I thought you wanted to keep them private, that is why I added the `getLogger()` method, but indeed using `logger` directly makes things easier here.
Providing a logger directly to the `create()` method is nice indeed!
However I had to find a way to change the verbosity in an easier manner than providing a new logger object, so I check if `logger` is a string, and if so it is considered as a winston log level instead.

I also merged with the last commit... But I may have done it wrong, since travis build still crashes for other reasons with node 5.10.1... :frowning:

File: src/phantom.js [hunk 6]
```diff
@@ -45,7 +36,36 @@ export default class Phantom {
         }
 
         let pathToShim = path.normalize(__dirname + '/shim.js');
-        logger.debug(`Starting ${phantomPath} ${args.concat([pathToShim]).join(' ')}`);
+
+        let loggerLevel = process.env.DEBUG === 'true' ? 'debug' : 'info';
+        if(typeof config.logger === 'object') {
```
Comment (amir20) [inline src/phantom.js]: I think you should be able to use es6's default params. Like 

`function create(params = [], {logger: defaultLogger, level: defaultLevel}`. I am pretty sure that would work but I have to try it.

File: src/phantom.js [hunk 7]
```diff
@@ -69,34 +89,34 @@ export default class Phantom {
...
                 }
             } else {
-                logger.info(message);
+                this.logger.info(message);
             }
         });
 
-        this.process.stderr.on('data', data => logger.error(data.toString('utf8')));
-        this.process.on('exit', code => logger.debug(`Child exited with code {${code}}`));
+        this.process.stderr.on('data', data => this.logger.error(data.toString('utf8')));
```
Comment (amir20) [inline src/phantom.js]: Anyway to add a test for all of this? I am guessing you can pass a mocked object now and assert that `error` gets called.

**Comment by amir20** [comment]
Good start. I added some comments.

**Comment by amir20** [comment]
I can also do some of this on Monday and merge.

File: src/phantom.js (hunk 6 above)
Comment (Dagrut) [inline src/phantom.js]: I tested it, it works.
However how to provide a default logger this way? Switching again to a global logger?

Also, if the default level is provided here, it would mean that the DEBUG environment variable would override its value, except if we remove it (which I think is a bad idea since it breaks the normal behavior), or if we put it outside of Phantom like logger was before and then use it here.

File: src/phantom.js (hunk 7 above)
Comment (Dagrut) [inline src/phantom.js]: Yes, I think that providing a custom phantom path that would do something like `echo "test" 1>&2` and then catching the output would be possible (Also I tested it now and it printed "test" as expected, with node 4.4.7).

However the syntax used in test files is not really familiar to me... I'll have a closer look at it later.

File: src/phantom.js (hunk 6 above)
Comment (amir20) [inline src/phantom.js]: Yea a global logger would be the way. 

Hmm, let me think about the `DEBUG` problem.  It makes sense for it to override.

Comment (amir20) [inline src/phantom.js]: Maybe we can put the logic for `DEBUG` in `index.js#create`

Comment (Dagrut) [inline src/phantom.js]: Okay, I think I see what you mean. However each time a `phantom` instance is created using the default logger with a different log level it would override the previous log level value (which is still better than before, when we had almost no way to change it at all :smile: ).

Comment (amir20) [inline src/phantom.js]: Yea, I didn't think about that. You are right. Hmm. I'd say let's just do it that way since it's all in one application.

**Comment by amir20** [comment]
Also, you have to rebase since I merged another PR.

File: src/phantom.js (hunk 6 above)
Comment (Dagrut) [inline src/phantom.js]: Okay! Also we can add `phantomPath` in the `create` call too, like `config = {phantomPath: phantomjs.path}`. I tested it, and it works too.

Comment (Dagrut) [inline src/phantom.js]: I turned the problem around, but if we put a default value for `config` inside `index.js#create`, we will have to check its value, and these values (like the logLevel) or the value of will probably have to be checked again in the `phantom` constructor, so... It may not help. As I understand it, it would be something like : 
```js
import Phantom from "./phantom";
import winston from "winston";

const logLevel = process.env.DEBUG === 'true' ? 'debug' : 'info';
const logger = new winston.Logger({
    transports: [
        new winston.transports.Console({
            level: logLevel,
            colorize: true
        })
    ]
});

/**
 * Retuns a new instance of Phantom class
 * @param args
 * @returns {Promise}
 */
module.exports.create = (args, config = {logger: logger, logLevel: logLevel}) => {
    if (config.logger === logger) {
        logger.transports.console.level = logLevel;
    }
    return new Promise(resolve => resolve(new Phantom(args, config)));
};
```

File: README.md [hunk 8]
```diff
@@ -103,6 +108,28 @@ It may be a good idea to register handlers to `SIGTERM` and `SIGINT` signals wit
...
+/* Set the log level to 'debug' after creating  */
+phantom.create().then(function(ph) {
+    ph.logger.transports.console.level = 'debug';
+    // use ph
+});
+
+/* Set your logger object directly in the create call */
+var log = console.log;
+var nolog = function() {};
+phantom.create([], { info: log, warn: log, debug: nolog, error: log }).then(function(ph) {
```
Comment (amir20) [inline README.md]: very nice

File: src/phantom.js [hunk 9]
```diff
@@ -45,7 +36,36 @@ export default class Phantom {
...
+                transports: [
+                    new winston.transports.Console({
+                        level: loggerLevel,
+                        colorize: true
+                    })
+                ]
+            });
+        }
+
+        this.logger.debug(`Starting ${phantomPath} ${args.concat([pathToShim]).join(' ')}`);
```
Comment (amir20) [inline src/phantom.js]: Do you still need all the code above? Ideally, we only change what is needed for this pull request.

**Comment by amir20** [comment]
@Dagrut I think this is almost ready to merge. I like your last suggestion on `create()`. But I didn't see the changes actually pushed. Also, is that huge block of code still needed? 

Let me know and thanks for being patient!

**Comment by Dagrut** [comment]
Okay!

That's right, I didn't push these changes (I didn't even save them), but it looked a little weird to me, and I didn't want to go back & forth because of a misunderstanding.

So here is the new version. I wanted to throw an error when logger is invalid in the Phantom constructor, but then most tests would fail, since they are calling the constructor itself, so I chose to fill missing methods with an empty function.

File: src/phantom.js [hunk 10]
```diff
@@ -45,7 +36,20 @@ export default class Phantom {
         }
 
         let pathToShim = path.normalize(__dirname + '/shim.js');
-        logger.debug(`Starting ${phantomPath} ${args.concat([pathToShim]).join(' ')}`);
+
+        if (typeof config.logger !== 'object') {
```
Comment (amir20) [inline src/phantom.js]: Why is this needed? Is it so that people might pass a bad logger?

File: src/index.js [hunk 11]
```diff
@@ -1,8 +1,31 @@
...
+ * @param [config] configuration object
+ * @param [config.phantomPath] path to phantomjs executable
+ * @param [config.logger] object containing functions used for logging
+ * @param [config.logLevel] log level to apply on the logger (if unset or default)
  * @returns {Promise}
  */
-module.exports.create = (args, config) => new Promise(resolve => resolve(new Phantom(args, config)));
+module.exports.create = (args, config = {logger: logger, logLevel: logLevel}) => {
+    if(!config.logger) {
+        config.logger = logger;
```
Comment (amir20) [inline src/index.js]: I think I prefer to simplify here. Instead of defaulting to a value, just throw exception if it is bad value. Because if someone is passing a bad value, then it is probably a bug. Better to call it out than assume intention.

File: src/phantom.js (hunk 10 above)
Comment (amir20) [inline src/phantom.js]: Same logic as above, maybe throw exception. Don't worry too much about cleaning up config.

File: src/phantom.js [hunk 12]
```diff
@@ -69,34 +73,34 @@ export default class Phantom {
...
         });
 
-        this.process.stderr.on('data', data => logger.error(data.toString('utf8')));
-        this.process.on('exit', code => logger.debug(`Child exited with code {${code}}`));
+        this.process.stderr.on('data', data => this.logger.error(data.toString('utf8')));
+        this.process.on('exit', code => this.logger.debug(`Child exited with code {${code}}`));
         this.process.on('error', error => {
-            logger.error(`Could not spawn [${phantomPath}] executable. Please make sure phantomjs is installed correctly.`);
-            logger.error(error);
+            this.logger.error(`Could not spawn [${phantomPath}] executable. Please make sure phantomjs is installed correctly.`);
```
Comment (amir20) [inline src/phantom.js]: Could you add a test? like a simple one where logger is working.

File: src/index.js [hunk 13]
```diff
@@ -1,8 +1,31 @@
...
+ * Retuns a Promise of a new Phantom class instance
+ * @param args command args to pass to phantom process
+ * @param [config] configuration object
+ * @param [config.phantomPath] path to phantomjs executable
+ * @param [config.logger] object containing functions used for logging
+ * @param [config.logLevel] log level to apply on the logger (if unset or default)
  * @returns {Promise}
  */
-module.exports.create = (args, config) => new Promise(resolve => resolve(new Phantom(args, config)));
+module.exports.create = (args, config = {logger: logger, logLevel: logLevel}) => {
```
Comment (amir20) [inline src/index.js]: 👍

File: src/phantom.js (hunk 10 above)
Comment (Dagrut) [inline src/phantom.js]: If I don't replace/fix the logger object, almost all tests are failing : 
```

File: src/index.js (hunk 11 above)
Comment (Dagrut) [inline src/index.js]: Es6's default arguments are passed as a whole, so if the user is sending `{}` or `{ phantomPath: "/some/path" }` as config, there will be no`logger` inside it, since it will take their object and not ours.

File: src/phantom.js (hunk 12 above)
Comment (Dagrut) [inline src/phantom.js]: Okay, I'll work on it!

File: src/phantom.js (hunk 10 above)
Comment (amir20) [inline src/phantom.js]: Oh crap. Well sounds like we should move the default to `new Phantom()` then. I'll fix that my self since you have been so patient. :)

File: src/index.js (hunk 11 above)
Comment (amir20) [inline src/index.js]: Yea, so originally when I was saying we should do this, I meant to use ES6's destructuring args. I wasn't sure if I had setup Babel to work that way though. If it works for you that is perfect! Let's do that.

File: src/phantom.js (hunk 7 above)
Comment (amir20) [inline src/phantom.js]: It's just using generators. So instead of doing `promise.then(...)` you can do `yield fromPromise()`.

**Comment by amir20** [comment]
I created a new PR at #535. I was trying to figure out how to keep your commits in it so that I don't take credit away but then as I was rebasing, they all disappeared. :( I am sorry. Thanks for the amazing job! Can you please review the other PR? With your help I can merge.

**Comment by amir20** [comment]
Closing in favor of #535.

**Comment by Dagrut** [comment]
Okay, thanks!
Don't worry about the authorship of the commits, it's fine to me :smiley: .
I'll check it now!
